### PR TITLE
Implement SetState

### PR DESCRIPTION
**Pull request by boaz-codota**

### CODE REVIEW

File: src/main/java/com/tabnine/contracts/AutocompleteRequest.java [hunk 1]
```diff
@@ -1,28 +1,36 @@
...
+
+public class AutocompleteRequest implements BinaryRequest<Map<String, Object>, AutocompleteResponse> {
     public String before;
     public String after;
     public String filename;
-    public boolean region_includes_beginning;
-    public boolean region_includes_end;
-    public int max_num_results;
-    public Integer correlation_id;
+    @SerializedName(value = "region_includes_beginning")
```
Comment (dimacodota) [inline src/main/java/com/tabnine/contracts/AutocompleteRequest.java:16]: nice 👍

File: src/main/java/com/tabnine/DependencyContainer.java [hunk 2]
```diff
@@ -2,10 +2,33 @@
...
+
+            GATEWAY_INSTANCE.init();
+        }
+
+        return GATEWAY_INSTANCE;
+    }
+
+    public static synchronized TabNineLookupListener singletonOfTabNineLookupListener() {
+        if (LOOKUP_LISTENER_INSTANCE == null) {
+            LOOKUP_LISTENER_INSTANCE = new TabNineLookupListener();
```
Comment (dimacodota) [inline src/main/java/com/tabnine/DependencyContainer.java:26]: why is it singleton? what state does it manages?

**Review by dimacodota — APPROVED** [review]
LGTM, had minor questin

File: src/main/java/com/tabnine/DependencyContainer.java (hunk 2 above)
Comment (boaz-codota) [reply src/main/java/com/tabnine/DependencyContainer.java:26]: Well they way IntelliJ does selections is weird, as it requires me to register the selection listener upon a completion suggestion (we did the same in Codota & I searched the web for how to do it), so to make sure that I won't register/initiate `TabNineLookupListener` multiple times I make it a singleton and removes it from the selection listener list before adding it again. At:
https://github.com/codota/tabnine-intellij/pull/53/files#diff-04d46f64b334bd2109e21f7780c6176fR133

Comment (dimacodota) [reply src/main/java/com/tabnine/DependencyContainer.java:26]: 👍

**Pull request merged**